Remove dynamic invocation

rrousselGit opened the pull request:
Since the matchers are using FeatureMatcher, the typedMatches method receives a typed object instead of dynamic

rrousselGit committed:
Remove dynamic invocation

Since the matchers are using FeatureMatcher, the typedMatches method receives a typed object instead of dynamic

rrousselGit committed:
Update num matchers

jakemac53 commented:
LGTM, can  you bump this to 0.12.15-dev and add a changelog entry? You can leave the message blank, since its not a user facing change.

natebosch committed:
Merge branch 'master' into patch-1

natebosch committed:
Enable lint avoid_dynamic_calls

Fix more cases, and ignore the intentional dynamic calls.
Add changelog

rrousselGit commented:
Ah sorry, I missed your previous message. Looks like the changelog entry + bump has been done already

natebosch committed:
Merge branch 'master' into patch-1

natebosch committed:
Change in type check output

natebosch committed:
Merge branch 'master' into patch-1

natebosch committed:
Move changelog entry

natebosch commented on lib/src/core_matchers.dart:
```diff
@@ -14,7 +14,9 @@ class _Empty extends Matcher {
   @override
-  bool matches(Object? item, Map matchState) => (item as dynamic).isEmpty;
+  bool matches(Object? item, Map matchState) =>
+      // ignore: avoid_dynamic_calls
```
It's too bad we don't have a better way to suppress these than the ignore. I filed https://github.com/dart-lang/linter/issues/4806

rrousselGit replied:
Could this be replaced with `if (item is Iterable) return item.isEmpty; else if (item is String) return item.isEmpty;`? 

That would technically be breaking if someone defined an unknown object with an `isEmpty` getter. But the fact that it currently works is surprising to me. I thought the implementation relied on pattern match, not dynamic invocation

natebosch replied:
I think supporting an `isEmpty` property on an arbitrary user defined class was an intentional feature and I'm not eager to try to remove it. It's not a design we'd carry forward to `checks`.

natebosch commented on lib/src/expect/throws_matcher.dart:
```diff
@@ -73,7 +73,7 @@ class Throws extends AsyncMatcher {
   @override
   dynamic /*FutureOr<String>*/ matchAsync(Object? item) {
-    if (item is! Function && item is! Future) {
+    if (item is! Function() && item is! Future) {
```
This stopped accepting values which are a `Function<T>()`. I think we may need to keep this as a dynamic call to handle generic methods. This is why the `Future.value` tear-off case is failing.

cc @jakemac53 - I have completely overlooked the fact that functions are not subtypes if they have different numbers of generic type arguments... I wonder if there are any other places we have been more strict than we should and are unintentionally blocking generic functions used as callbacks.

natebosch committed:
Retain a check against `Function`

The refactoring started rejecting generic functions, a `Function<T>()`
is not a subtype of a `Function()`.

natebosch commented on lib/src/core_matchers.dart:
```diff
@@ -144,11 +148,11 @@ class isInstanceOf<T> extends TypeMatcher<T> {
 
   @override
-  bool typedMatches(Function f, Map matchState) {
+  bool typedMatches(Function() f, Map matchState) {
```
This is likely a behavior change that we don't have tests for...

natebosch committed:
More places where Function() is too strict
